apps/api/src/app.gateway.ts

```diff
@@ -0,0 +1,106 @@
+import {
+  ConnectedSocket,
+  MessageBody,
+  SubscribeMessage,
+  WebSocketGateway,
+  WebSocketServer,
+  WsResponse,
+} from '@nestjs/websockets';
+import { Server, Socket } from 'socket.io';
+import {
+  CACHE_MANAGER,
+  Inject,
+  Logger,
+  UnauthorizedException,
+  UseFilters,
+} from '@nestjs/common';
+import { Cache } from 'cache-manager';
+import { AuthService } from './auth/auth.service';
+import { HttpExceptionTransformationFilter } from './common/filters/HttpExceptionFilter.filter';
+import { SpiedUserDto } from './spied-user.dto';
+import { User } from 'types';
+
+@WebSocketGateway()
+@UseFilters(HttpExceptionTransformationFilter)
+export class AppGateway {
+  @WebSocketServer() server!: Server;
+
+  constructor(
+    private readonly logger: Logger,
+    @Inject(CACHE_MANAGER) private readonly cacheManager: Cache,
+    private readonly authService: AuthService,
+  ) {}
+
+  async handleConnection(client: Socket): Promise<void> {
+    if (!client.handshake.auth.token)
+      throw new UnauthorizedException('No Authorization header found');
+    const currentUser = this.authService.verifyUserFromToken(
+      client.handshake.auth.token,
+    );
+    if (!currentUser) throw new UnauthorizedException('Invalid token');
+    client.data = { id: currentUser.sub, name: currentUser.name };
+    client.join(`user_${currentUser.sub}`);
+    this.server.to(`user_${currentUser.sub}`).emit('user_status_update', {
+      type: 'online',
+      userId: client.data.id,
+    });
+    this.cacheManager.set(`user:${currentUser.sub}`, currentUser, 0);
+    this.logger.log(
+      `'${currentUser.sub}' (${currentUser.name}) connected on socket '${client.id}'`,
+    );
+  }
+
+  async handleDisconnect(client: Socket): Promise<void> {
+    await this.server
+      .in(`user_${client.data.id}`)
+      .fetchSockets()
+      .then((sockets) => {
+        if (
+          sockets.filter((socket) => socket.data.id === client.data.id)
+            .length === 0
+        ) {
+          this.cacheManager.del(`user:${client.data.id}`);
+          this.server.to(`user_${client.data.id}`).emit('user_status_update', {
+            type: 'offline',
+            userId: client.data.id,
+          });
+          this.logger.log(
+            `'${client.data.id}' (${client.data.name}) disconnected from socket '${client.id}'`,
+          );
+        }
+      });
+  }
+
+  @SubscribeMessage('subscribe_user')
+  async subscribeUser(
+    @ConnectedSocket() client: Socket,
+    @MessageBody() spiedUserDto: SpiedUserDto,
+  ): Promise<WsResponse<unknown>> {
+    this.logger.log(
+      `'${client.data.id}' (${client.data.name}) is spying on '${spiedUserDto.userId}'`,
+    );
+    client.join(`user_${spiedUserDto.userId}`);
+    const user = await this.cacheManager.get<User>(
+      `user:${spiedUserDto.userId}`,
+    );
+    return {
```

Comment:
You can also simply return the data from you function, this will then be available in the callback of the emit method in the front. Up to you!

Reply:
I've added another function to prevent code duplication. Now the subscription has the same callback as the event handler.